File: app/utils/events/easyEvents.lua
local ____lualib = require("lualib_bundle")
local __TS__Spread = ____lualib.__TS__Spread
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
function ____exports.newEasyEvents(events, callbacksDeclare)
    local _callbackCall
    function _callbackCall(listener, args)
        local isOk = true
        local err = ""
        local ____listener_context_2 = listener.context
        if ____listener_context_2 == nil then
            ____listener_context_2 = {}
        end
        local context = ____listener_context_2
        local res = {pcall(function()
            listener.callback(
                context,
                __TS__Spread(args)
            )
        end)}
        isOk = res[1]
        err = res[2]
        return {isOk = isOk, err = err}
    end
    local listeners = {}
    for ____, eventKey in ipairs(events) do
        listeners[eventKey] = {}
    end
    local function on(event, callback, context)
        if listeners[event] == undefined then
            listeners[event] = {}
        end
        local ____listeners_event_0 = listeners[event]
        ____listeners_event_0[#____listeners_event_0 + 1] = {callback = callback, context = context}
    end
    local function emit(event, ...)
        local args = {...}
        if listeners[event] ~= undefined then
            for ____, listener in ipairs(listeners[event]) do
                local ____callbackCall_result_1 = _callbackCall(listener, args)
                local isOk = ____callbackCall_result_1.isOk
                local err = ____callbackCall_result_1.err
                if not isOk then
                    Log.error((("Emite event " .. event) .. " error: ") .. tostring(err))
                end
            end
        end
    end
    local function final()
        for ____, key in ipairs(events) do
            local listenerList = listeners[key]
            if listenerList ~= undefined and #listenerList > 0 then
                __TS__ArraySplice(listenerList, 0, #listenerList - 1)
            end
            __TS__Delete(listeners, key)
        end
    end
    return {on = on, emit = emit, final = final}
end
return ____exports

File: src/utils/events/easyEvents.ts
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
export type EventCallbacks<Events extends readonly string[]> = { [key in Events[number]]: (this: any, ...args: any[]) => void };
/**@noSelf */
export interface EasyEvents<Event extends readonly string[], Callbacks extends EventCallbacks<Event>> {
    on<T extends Event[number]>(event: T, callback: Callbacks[T], context?: any): void;
    emit<T extends Event[number]>(event: T, ...args: Parameters<Callbacks[T]>): void;
    final(): void;
}

export function newEasyEvents<Event extends readonly string[], C extends EventCallbacks<Event>>
    (events: Event, callbacksDeclare: C): EasyEvents<Event, C> {


    type Listeners = { [event: string]: Array<{ callback: C[Event[number]], context?: any }> };

    const listeners: Listeners = {} as Listeners;

    for (const eventKey of events)
        listeners[eventKey] = [];

    function on<T extends Event[number]>(event: T, callback: C[T], context?: any): void {
        if (listeners[event as string] == undefined) listeners[event as string] = [];
        listeners[event as string].push({ callback, context });
    }

    function emit<T extends Event[number]>(event: T, ...args: Parameters<C[T]>): void {
        if (listeners[event as string] != undefined)
            for (const listener of listeners[event as string]) {
                const { isOk, err } = _callbackCall(listener, args);
                if (!isOk)
                    Log.error('Emite event ' + (event as string) + ' error: ' + err);
            }
    }

    function _callbackCall<T extends Event[number]>(listener: { callback: C[T], context?: any }, args: Parameters<C[T]>): { isOk: boolean, err: any } {
        let isOk = true;
        let err: any = '';
        const context = listener.context ?? {};

        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        // pcall(listener.callback, {}, args);
        const res = pcall(() => {
            listener.callback.call(context, ...args);
        });

        isOk = res[0];
        err = res[1];

        return { isOk, err };
    }

    function final() {
        for (const key of events) {
            const listenerList = listeners[key];
            if (listenerList != undefined && listenerList.length > 0)
                listenerList.splice(0, listenerList.length - 1);

            delete listeners[key];
        }  
    }

    return { on, emit, final };
}